src/middlewared/middlewared/plugins/asigra.py

```diff
@@ -0,0 +1,358 @@
+
+        except Exception as e:
+            self.logger.error("Can't connect to dssystem: {}".format(e))
+            return False
+
+        con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+        cur = con.cursor()
+
+        files = glob.glob("{}/dssp*.sql".format(self.dssystem_db_path))
+
+        max = 0
+        for f in files:
+            m = re.match('.*/dssp([0-9]+).sql', f)
+            if not m or len(m.groups()) != 1:
+                continue
+            if int(m.group(1)) > max:
+                max = int(m.group(1))
+
+        db_number = await self.get_db_number()
+        if db_number < 0:
+            db_number *= -1
+
+        db_number += 1
+        while max >= db_number:
+            sql_patch = os.path.join(self.dssystem_db_path, "dssp{}.sql".format(db_number))
+
+            self.logger.debug("apply the patch dssp{}.sql".format(db_number))
+            with open(sql_patch, "r") as f:
+                try:
+                    cur.execute(f.read())
+
+                except Exception as e:
+                    self.logger.error("Can't init dssystem database: {}".format(e))
+                    return False
+
+            db_number += 1
+
+        cur.close()
+        con.close()
+
+        return True
+
+    async def install_asigra(self):
+        ret = True
+
+        if os.path.exists("/asigra"):
+            dssystem_pkg = None
+
+            files = glob.glob("/asigra/DS-System_*/dssystem-*.txz")
+            for f in files:
+                dssystem_pkg = f
+
+            if dssystem_pkg:
+                with open("/tmp/dssystem_install.ini", "w") as f:
+                    f.write("silentmode=yes\n")
+
+                proc = await Popen(
+                    ['/usr/sbin/pkg', 'add', '-f', dssystem_pkg],
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.PIPE,
+                    close_fds=True,
+                    env={'INSTALL_INI': '/tmp/dssystem_install.ini'}
+                )
+            else:
+                ret = False
+
+            output = (await proc.communicate())[0].decode()
+            if proc.returncode != 0:
+                self.logger.error(output)
+                self.logger.error('Failed to install dssystem package:\n{}'.format(output))
+                ret = False
+
+            if not ret:
+                try:
+                    os.chmod("/usr/local/etc/rc.d/dssystem", 0o555)
+                    shutil.copyfile("/usr/local/etc/rc.d/dssystem", "/conf/base/etc/local/rc.d/dssystem")
+                    os.chmod("/conf/base/etc/local/rc.d/dssystem", 0o555)
+
+                except Exception:
+                    self.logger.error("Failed to install dssystem rc.d script")
+                    ret = False
+
+        return ret
+
+    async def asigra_installed(self):
+        if not os.path.exists(os.path.join(self.dssystem_path, "libasigraencmodule.so")):
+            return False
+        return True
+
+    async def setup_asigra(self):
```

Comment:
Given that almost all the methods used here should not be a coroutine, it makes sense for this one to be a normal method as well.